File: LeetCode/Python/IntegerToRoman.py
__author__ = 'YiLIU'
class Solution:
    def __init__(self):
        self.units = ["I", "V", "X"]
        self.tens = ["X", "L", "C"]
        self.hundreds = ["C", "D", "M"]
        self.result = ""

    def intToRoman(self, num):
        if num >= 1000:
            digit = num / 1000
            num -= digit * 1000
            self.result = "M" * digit
            self.result += self.intToRoman(num)
        elif num >= 100:
            digit = num / 100
            num -= digit * 100
            self.result = self.intToRomanHelper(digit, self.hundreds)
            self.result += self.intToRoman(num)
        elif num >= 10:
            digit = num /10;
            num -= digit * 10
            self.result = self.intToRomanHelper(digit, self.tens)
            self.result += self.intToRoman(num)
        else:
            self.result = self.intToRomanHelper(num, self.units)
        return self.result

    def intToRomanHelper(self, digit, level):
        if digit >= 9:
            return level[0] * (10 - digit) + level[2]
        elif digit >= 5:
            return level[1] + level[0] * (digit - 5)
        elif digit >= 4:
            return level[0] * (digit - 3) + level[1]
        else:
            return level[0] * digit

print Solution().intToRoman(2 ** 31 - 1)
print Solution().intToRoman(1990)
print Solution().intToRoman(2014)

File: LeetCode/Java/IntegerToRoman.java
/**
 * Created by yi on 5/7/15.
 */
public class IntegerToRoman {
    private static final int[] values = {1, 5, 10, 50, 100, 500, 1000};
    private static final Character[] romans = {'I', 'V', 'X', 'L', 'C', 'D', 'M'};

    public static void main(String[] args) {
        System.out.println(new IntegerToRoman().intToRoman(45));
    }

    public String intToRoman(int num) {
        StringBuilder result = new StringBuilder();
        for (int i = 6; i >= 0; --i) {
            int n = num / values[i];
            for (int j = 0; j < n; ++j)
                result.append(romans[i]);
            num -= n * values[i];
            int validMinusIndex = 0;
            if (i > 4) validMinusIndex = 4;
            else if (i > 2) validMinusIndex = 2;
            else validMinusIndex = 0;
            if (i > 0 && values[i] / values[validMinusIndex] - num / values[validMinusIndex] == 1) {
                result.append(romans[validMinusIndex]);
                result.append(romans[i]);
                num -= (num / values[validMinusIndex]) * values[validMinusIndex];
            }
        }
        return result.toString();
    }
}
